packages/appsync-modelgen-plugin/src/utils/process-has-many.ts

```diff
@@ -134,3 +135,72 @@ export function getConnectedFieldsForHasMany(
+ */
+ function addKeyToModel(model: CodeGenModel, name: string, fields: string[]): void {
+  model.directives.push({
+    name: 'key',
+    arguments: {
+      name,
+      fields,
+    },
+  });
+}
+
+/**
+ * Helper to simplify retrieving either CPK-enabled or standard connection fields.
+ * Returns a list of fields of at least length 1.
+ */
+function getConnectionAssociatedFields(hasManyConnection: CodeGenFieldConnectionHasMany): CodeGenField[] {
+  const associatedFields = hasManyConnection.associatedWithFields && hasManyConnection.associatedWithFields.length > 0
+  ? hasManyConnection.associatedWithFields
+  : [hasManyConnection.associatedWith]
+  if (associatedFields.length === 0) {
+    throw new Error('Expected at least one associated field for the hasMany relationship.');
+  }
+  return associatedFields;
+}
+
+/**
+ * A corresponding belongsTo means that the corresponding model has a field with type that is the name of this model, and with a @belongsTo
+ * directive attached to it.
+ */
+function doesHasManyConnectionHaveCorrespondingBelongsTo(model: CodeGenModel, hasManyConnection: CodeGenFieldConnectionHasMany): boolean {
+  const fieldReferencingParent = hasManyConnection.connectedModel.fields.find(f => f.type === model.name);
+  if (!fieldReferencingParent) {
+    return false;
+  }
+  return fieldReferencingParent.directives.some(d => d.name === TransformerV2DirectiveName.BELONGS_TO);
+}
+
+/**
+ * This is a bit backwards, but we need to check if this connection directive specifies an index.
+ */
+function doesHasManySpecifyIndexName(field: CodeGenField): boolean {
+  return field.directives.some(d => d.name === TransformerV2DirectiveName.HAS_MANY && d.arguments.indexName);
+}
+
+/**
+ * Return whether or not a hasMany connection has an implicit key defined.
+ * This is determined if there is a belongsTo directive on the connected model, or if there is an index defined on the directive.
+ */
+export function hasManyHasImplicitKey(field: CodeGenField, model: CodeGenModel, hasManyConnection: CodeGenFieldConnectionHasMany): boolean {
+  const hasCorrespondingBelongsTo = doesHasManyConnectionHaveCorrespondingBelongsTo(model, hasManyConnection);
+  const hasIndexNameSpecified = doesHasManySpecifyIndexName(field);
+  return !(hasCorrespondingBelongsTo || hasIndexNameSpecified);
+}
+
+/**
+ * Extract the name and list of keys from the connection information, and add the key to
+ * the related model.
+ */
+export function addHasManyKey(hasManyConnection: CodeGenFieldConnectionHasMany): void {
+  const associatedFieldNames = getConnectionAssociatedFields(hasManyConnection).map(f => f.name);
+  const connectedModel = hasManyConnection.connectedModel;
+  // Applying consistent auto-naming as the transformer does today
+  // https://github.com/aws-amplify/amplify-category-api/blob/main/packages/amplify-graphql-relational-transformer/src/resolvers.ts#L334-L396
+  const name = `gsi-${connectedModel.name}.${associatedFieldNames[0]}`;
```

Comment:
I just test locally and find that this naming in transformer should be `gsi-${parent-model.name}.${hasManyFieldName}`
Take the testing schema as example
```GraphQL
type Project @model {
  id: ID! @primaryKey(sortKeyFields: ["name"])
  name: String!
  teams: [Team] @hasMany
}
type Team @model {
  name: String!
}
```
The GSI name is `gsi-Project.teams`. (Confirmed with the `build/stacks/Team.json` in the api folder)

Reply:
The CPK feature flag also changes a behavior for the auto-generated index. When the feature flag is true, it will add  all the associated fields in the index component(aka `projectTeamsId` and `projectTeamsName` in the example). When it is false, it only uses the primary key field(`projectTeamId`) for the index (no sort key fields defined). 
I notice `getConnectionAssociatedFields` uses the same logic. We can add a unit test to verify this scenario.

Reply:
I actually do have a case that flexes that behavior, in the `process-has-many` tests, the `works with non-cpk fallback connection` case should cover this, by providing an empty associatedWithFields list. I'll verify I can pass a null in as well.